### PR TITLE
GH-8220: Approvals e2e tests ( regular & accessibility )

### LINKED ISSUE #8220: End to end tests for approval dashboard
Needs to test
- cancel functionality
- reject functionality
- approve functionality 

as approver

* note make sure each `it` are idempotent and does not rely to previous state. Meaning each `it` should pass without depending on the previous state.

### PR DESCRIPTION
**Pull request by plabadie**
closes GH-8220

### CODE REVIEW

File: projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-order-approval.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+  );
+
+  cy.visit(`/my-account/approval/${sampleData.approvalOrderDetail.code}`);
+  cy.wait(`@${alias}`).its('status').should('eq', 200);
+}
+
+export function loginB2bUser() {
+  const user: AccountData = {
+    user: '',
+    registrationData: {
```
Comment (hackergil) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-order-approval.ts]: This is duplicated below (except for one line), and also there are similar snippets in other helper files. Can we get rid of the duplications as much as possible?

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/accessibility/tabbing-order.e2e-spec.ts [hunk 2]
```diff
@@ -131,4 +141,39 @@ describe('Tabbing order for B2B checkout', () => {
       });
     });
   });
+
+  describe('Order Approval', () => {
+    before(() => {
+      cy.window().then((win) => win.sessionStorage.clear());
+      Cypress.env('BASE_SITE', POWERTOOLS_BASESITE);
```
Comment (hackergil) [inline projects/storefrontapp-e2e-cypress/cypress/integration/b2b/accessibility/tabbing-order.e2e-spec.ts]: This is already a B2B test. Can we find a way to not set the baseSite every time?

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/accessibility/tabbing-order.e2e-spec.ts [hunk 3]
```diff
@@ -131,4 +141,39 @@ describe('Tabbing order for B2B checkout', () => {
...
+    });
+
+    context('Approval List', () => {
+      it('should allow to navigate with tab key', () => {
+        loginB2bApprover();
```
Comment (hackergil) [inline projects/storefrontapp-e2e-cypress/cypress/integration/b2b/accessibility/tabbing-order.e2e-spec.ts]: Duplication on calling loginB2BApprover can be simplified placing it on a beforeEach() hook. Or even better, just once on before() hook.

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+      .eq(5)
+      .should(
+        'contain',
+        sampleData.approvalOrderList.orderApprovals[0].order.totalPrice
+          .formattedValue
+      );
+  });
+
+  it('Should display approval detail page', () => {
+    orderApproval.loginB2bApprover();
```
Comment (hackergil) [inline projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts]: Same as above. login can be done on beforeEach() or before() hook

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+  // assert order status
+  cy.get('cx-order-details-items .cx-list-status').should(
+    'contain',
+    sampleData.statusPendingApproval
+  );
+
+  // assert products
+  cy.get('.cx-item-list-row .cx-link').should(
+    'contain',
+    sampleData.approvalOrderDetail.order.entries[0].product.name
```
Comment (hackergil) [inline projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts]: Can we shorten this to
firstOrderEntry = sampleData.approvalOrderDetail.order.entries[0]
and then assert on firstOrderEntry.* ?
(law of Demeter)

File: projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-order-approval.ts [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+      password: '12341234',
+      titleCode: 'mr',
+      email: b2bApprover.uid,
+    },
+  };
+  cy.requireLoggedIn(approver);
+  visitHomePage();
+}
+
+export function getPendingOrderDetails() {
```
Comment (hackergil) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-order-approval.ts]: can we make more explicit in the name of these functions that we are stubbing the response of an xhr call?

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+    .eq(3)
+    .should(
+      'contain',
+      `${sampleData.approvalOrderDetail.order.deliveryAddress.firstName} ${sampleData.approvalOrderDetail.order.deliveryAddress.lastName}`
+    );
+  cy.get('cx-order-overview cx-card')
+    .eq(3)
+    .should(
+      'contain',
+      sampleData.approvalOrderDetail.order.deliveryAddress.formattedAddress
```
Comment (hackergil) [inline projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts]: Same comment as above

**Review by hackergil — CHANGES_REQUESTED** [review]
Some questions:
* I had to update my cypress file to start as b2b with the right basesite. What's the plan to handle this? Do we know?
* What happened with the idea of incorporating tabbing to the test?
* I noticed on the order approval list page there are some dropdowns that are not being tested. Why?

File: projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-order-approval.ts (hunk 6 above)
Comment (plabadie) [reply projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-order-approval.ts]: Done. We update the names of all functions which use a stubbed response.

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/accessibility/tabbing-order.e2e-spec.ts (hunk 2 above)
Comment (WeizhengSap) [reply projects/storefrontapp-e2e-cypress/cypress/integration/b2b/accessibility/tabbing-order.e2e-spec.ts]: done, `Cypress.env('BASE_SITE', POWERTOOLS_BASESITE)` was removed

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/accessibility/tabbing-order.e2e-spec.ts (hunk 3 above)
Comment (WeizhengSap) [reply projects/storefrontapp-e2e-cypress/cypress/integration/b2b/accessibility/tabbing-order.e2e-spec.ts]: done, moved to `beforeEach()`.

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts (hunk 4 above)
Comment (WeizhengSap) [reply projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts]: done, move to `beforeEach()`

**Comment by plabadie** [comment]
> Some questions:
> 
> * I had to update my cypress file to start as b2b with the right basesite. What's the plan to handle this? Do we know?

The yarn scripts provide the base site as an argument when they start the b2b tests.
Locally you can use some of the existing commands in the package.json or add "--env BASE_SITE=powertools-spa" to the commands you use to launch cypress. 

> * What happened with the idea of incorporating tabbing to the test?

We had already pushed this ticket for review when we had the team discussion, but we now incorporated the accessibility tests into the "regular" approvals e2e.

> * I noticed on the order approval list page there are some dropdowns that are not being tested. Why?

It is a shared component that drives this functionality, so I guess we left it out initially.  However, we now added coverage for the sorting dropdown.

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+import * as orderApproval from '../../../helpers/b2b/b2b-order-approval';
+import * as sampleData from '../../../sample-data/b2b-order-approval';
+
+describe('B2B - Order Approval', () => {
+  before(() => {
+    cy.window().then((win) => win.sessionStorage.clear());
+  });
+
+  describe('Order Approval - Accessibility.', () => {
+    beforeEach(() => {
```
Comment (hackergil) [inline projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts]: Logging in consumes some seconds. The pattern that we are following in other tests is:
- Log in once
- Store the token in local storage
- restore local storage before each test, to get the token
That way, you don't have to login again. You'll reduce the execution time for at least 30 seconds by logging in once or twice as an approver instead of 8 times (before each test)

```suggestion
    before(() => {
      orderApproval.loginB2bApprover();
      cy.saveLocalStorage();
    });

    beforeEach(() => {
      cy.restoreLocalStorage();
    });
```

File: projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-order-approval.ts [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    'approvalDetailPage'
+  );
+
+  cy.visit(`/my-account/approval/${sampleData.approvalOrderDetail.code}`);
+  cy.wait(`@${alias}`).its('status').should('eq', 200);
+}
+
+export function loginB2bUser() {
+  cy.requireLoggedIn(sampleData.b2bUserAccount);
+  visitHomePage();
```
Comment (hackergil) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-order-approval.ts]: I think you don't need to visit the homepage here. This takes about 3 seconds.

File: projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-order-approval.ts [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+export function loginB2bApprover() {
+  cy.requireLoggedIn(sampleData.b2bApproverAccount);
+  visitHomePage();
```
Comment (hackergil) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-order-approval.ts]: I think you don't need to visit the homepage here. This takes about 3 seconds.

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+      orderApproval.loginB2bApprover();
+    });
+
+    it('should display order approval list', () => {
+      orderApproval.getStubbedOrderApprovalList();
+      orderApproval.visitOrderApprovalListPage();
+      cy.wait('@order_approval_list').its('status').should('eq', 200);
+
+      cy.get('cx-order-approval-list a.cx-order-approval-value')
```
Comment (hackergil) [inline projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts]: Maybe there is a way to abstract these selectors into either a function or a block so that we don't repeat them?

**Review by hackergil — CHANGES_REQUESTED** [review]

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts [hunk 12]
```diff
@@ -0,0 +1,275 @@
+import { tabbingOrderConfig as config } from '../../../helpers/accessibility/b2b/tabbing-order.config';
+import * as orderApproval from '../../../helpers/b2b/b2b-order-approval';
+import * as sampleData from '../../../sample-data/b2b-order-approval';
+
+describe('B2B - Order Approval', () => {
+  before(() => {
+    cy.window().then((win) => win.sessionStorage.clear());
+  });
+
+  describe('Order Approval - Accessibility.', () => {
```
Comment (hackergil) [inline projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts]: My idea of including the accessibility tests within the main test was to have one test only that validated both the user tabbing and also the functionality (not to place the two tests/journeys in a single file). I checked how the tabbing helpers work and at the moment it is not possible. I apologize for not explaining my intention clearly.

Because of this, I think it's better if the accessibility tests are placed under b2b/accessibility on their own file.

File: projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-order-approval.ts (hunk 9 above)
Comment (plabadie) [reply projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-order-approval.ts]: Done.

File: projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-order-approval.ts (hunk 10 above)
Comment (plabadie) [reply projects/storefrontapp-e2e-cypress/cypress/helpers/b2b/b2b-order-approval.ts]: Done.

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts (hunk 8 above)
Comment (plabadie) [reply projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts]: Done.

File: projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts (hunk 11 above)
Comment (plabadie) [reply projects/storefrontapp-e2e-cypress/cypress/integration/b2b/order-approval/b2b-order-approval.e2e-spec.ts]: Done.

**Review by hackergil — APPROVED** [review]
